### PR TITLE
Add a Rust library

### PR DESCRIPTION
**Pull request by burgerdev**
More or less the first thing I did when I started this branch was researching the appropriate X25519 library to use. Now, almost 3 years later, I still find it rather hard to decide. My top picks were `x25519_dalek`, `ed25519_compact` and `openssl` bindings, in order of preference. But when I started with the Dalek crate, I encountered strange problems when using its `getrandom` feature, so I rewrote the thing for `ed25519_compact`. Quite a lot seems to have happened since then - I did not see any of the issues I had back then when I re-rewrote the thing just now. 

This train of rewrites is the main motivation behind me implementing two versions of the API. I really wanted to have an interface that does not leak the underlying X25519 library - this is how we ended up with the `untyped` module. Happy to hear your thoughts, but I, for one, consider 
```
fn tag(ours: &[u8; 32], theirs: &[u8; 32], ctr: u8, msg: &[u8]) -> [u8; 32]
```
... not as an improvement over C, so I wrote the `typed` module, too, which uses dalek types.

There is probably a third option, with an API that accepts traits and next to it a few impls for x25519 choices, but I don't think my Rust is up to that yet.

**EDIT - June 2024**: I implemented tagging and verification in terms of traits and provided trait impls for openssl and dalek. This is probably the way to go.

### CODE REVIEW

**Comment by vvidic** [comment]
Looks interesting, but my Rust is at a hello world level, so probably can't do a good review here :)

**Review by pkern — COMMENTED** [review]
So I agree that the untyped variant is not great and neither is the typed variant. Is there a way where we could offer our own wrapper types? Or could we export these types as type aliases somehow? In general the underlying types are very simple, the traits are where it's at...

**Comment by pkern** [comment]
This looks straightforward and simple now. Thank you! :D

I had one question on chat if you can now implement these traits from another crate. And two things I stumbled upon was the dalek import from lib and lib being no_std. But I'd be happy to merge this as-is. The PR is still technically in draft mode, though. (And I understand that there are the TODOs in the code, so maybe we still want to extend TagArgs a little bit before merging?

File: rust/examples/glome-cli.rs
Comment (burgerdev) [reply rust/examples/glome-cli.rs]: I removed the examples dir from this PR.

**Comment by burgerdev** [comment]
This should be good to review now, I'll add a Github Action later on.

**Comment by burgerdev** [comment]
@nicmue could you please take a look at this PR, with particular attention to module structure, Rust best practices and the test implementation? I wonder if there's a better way to execute all test vectors for a given `PublicKey`/`PrivateKey` implementation.

File: rust/src/lib.rs
Comment (nicmue) [inline rust/src/lib.rs:194]: `cargo clippy` warnings:
```suggestion
        assert!(verify(
            &apriv,
            &bpub,
            100,
            b"The quick brown fox",
            &hex!("06476f1f314b")
        ));
        assert!(!verify(
            &apriv,
            &bpub,
            100,
            b"The quick brown fox",
            &hex!("ffeeddccbbaa")
        ));
        assert!(!verify(
            &bpriv,
            &apub,
            100,
            b"The quick brown fox",
            &hex!("06476f1f314b")
        ));
```

File: rust/src/lib.rs
Comment (nicmue) [inline rust/src/lib.rs:62]: `cargo clippy` warning:

```suggestion
        .chain_update([ctr])
```

File: rust/src/openssl.rs
Comment (nicmue) [inline rust/src/openssl.rs:69]: `cargo clippy` warning:
```suggestion
        let public = secret.public_key();
```

File: rust/src/openssl.rs
Comment (nicmue) [inline rust/src/openssl.rs:47]: Yes! You can use assert everywhere. It will just panic if the check fails and I guess this is exactly what you want here.

File: rust/src/lib.rs
Comment (nicmue) [inline rust/src/lib.rs:113]: Since this is just a test it's not a must and more personal preference but you can use a generic type here:

```suggestion
    pub fn run_vector_1<T, F>(load_keypair: &F)
    where
        T: PrivateKey,
        F: Fn([u8; 32]) -> (T, T::PublicKey),
    {
```
Of course the same goes for `run_vector_2`.

File: rust/src/lib.rs
Comment (nicmue) [inline rust/src/lib.rs:77]:
```suggestion
/// The return value is `true` if the given tag matches the expected tag.
```

**Review by nicmue — COMMENTED** [review]
Since I am not familiar with the repo and the usage of this code I can't judge the functionality and therefore only the pure rust code: 

Looks really good from my side. I just pointed out the warnings you will get when running `cargo clippy` (which I highly recommend) and a personal preference on the `&dyn Fn...` call in your `run_vector` functions. The model structure and implementation looks completely idiomatic to me and tbh I like how you handled the test execution for all your `PublicKey`/`PrivateKey` implementation.

File: rust/src/dalek.rs
Comment (pkern) [inline rust/src/dalek.rs:13]: Can we make the verify take the actual tag result from tag, i.e. make this into an example that may roughly pass? Maybe bonus points to generate a key rather than picking 15 and 16? That 15 made my eyes twitch.

File: rust/src/openssl.rs
Comment (pkern) [inline rust/src/openssl.rs:17]: Same comment in spirit here.

**Review by pkern — CHANGES_REQUESTED** [review]
I like. Happy to merge once the klippy fixes are in.

File: rust/src/lib.rs
Comment (burgerdev) [reply rust/src/lib.rs:77]: Spelled it out.

File: rust/src/openssl.rs
Comment (burgerdev) [reply rust/src/openssl.rs:47]: Thanks!

File: rust/src/dalek.rs
Comment (burgerdev) [reply rust/src/dalek.rs]: Changed to the test vectors and verifying the generated tag. Not sure if I want to randomly generate a key in the doctests...

**Comment by burgerdev** [comment]
@nicmue Thanks so much for the review, and sorry for taking some time to get back to it. This was really helpful, and now I know about `clippy` (and `check`, for that matter). 

A follow-up to this PR should add these checks to CI.

**Comment by burgerdev** [comment]
@nicmue since you are now co-author here, do you want to sign Google's CLA? Alternatively, I can remove the `Co-authored-by` lines from the commits if you'd prefer.

**Comment by nicmue** [comment]
> @nicmue since you are now co-author here, do you want to sign Google's CLA? Alternatively, I can remove the `Co-authored-by` lines from the commits if you'd prefer.

Sure why not. What do I need to do? :D

**Comment by burgerdev** [comment]
The CLA bot check failure has some info: https://github.com/google/glome/pull/175/checks?check_run_id=26263501184. tl;dr: sign in to https://cla.developers.google.com/clas and agree to the terms.

**Comment by nicmue** [comment]
> The CLA bot check failure has some info: https://github.com/google/glome/pull/175/checks?check_run_id=26263501184. tl;dr: sign in to https://cla.developers.google.com/clas and agree to the terms.

Done you can try to retrigger. :)

**Review by pkern — APPROVED** [review]
LGTM, thank  you two very much!

**Pull request merged**
